cmd/bosun/search/search.go

```diff
@@ -24,51 +24,52 @@ import (
 type Search struct {
 	DataAccess database.DataAccess
 
-	last map[string]*lastInfo
+	// metric -> tags -> struct
+	last map[string]map[string]*database.LastInfo
 
 	indexQueue chan *opentsdb.DataPoint
 	sync.RWMutex
 }
 
-type lastInfo struct {
-	lastVal      float64
-	diffFromPrev float64
-	timestamp    int64
-}
-
 func init() {
 	metadata.AddMetricMeta("bosun.search.index_queue", metadata.Gauge, metadata.Count, "Number of datapoints queued for indexing to redis")
 	metadata.AddMetricMeta("bosun.search.dropped", metadata.Counter, metadata.Count, "Number of datapoints discarded without being saved to redis")
 }
+
 func NewSearch(data database.DataAccess) *Search {
 	s := Search{
 		DataAccess: data,
-		last:       make(map[string]*lastInfo),
+		last:       make(map[string]map[string]*database.LastInfo),
 		indexQueue: make(chan *opentsdb.DataPoint, 300000),
 	}
 	collect.Set("search.index_queue", opentsdb.TagSet{}, func() interface{} { return len(s.indexQueue) })
+	s.loadLast()
 	go s.redisIndex(s.indexQueue)
+	go s.backupLoop()
 	return &s
 }
 
 func (s *Search) Index(mdp opentsdb.MultiDataPoint) {
 	for _, dp := range mdp {
 		s.Lock()
-		metric := dp.Metric
-		key := metric + dp.Tags.String()
-		p := s.last[key]
+		mmap := s.last[dp.Metric]
+		if mmap == nil {
+			mmap = make(map[string]*database.LastInfo)
+			s.last[dp.Metric] = mmap
+		}
+		p := mmap[dp.Tags.String()]
 		if p == nil {
-			p = &lastInfo{}
-			s.last[key] = p
+			p = &database.LastInfo{}
+			mmap[dp.Tags.String()] = p
 		}
-		if p.timestamp < dp.Timestamp {
+		if p.Timestamp < dp.Timestamp {
 			if fv, err := getFloat(dp.Value); err == nil {
-				p.diffFromPrev = (fv - p.lastVal) / float64(dp.Timestamp-p.timestamp)
-				p.lastVal = fv
+				p.DiffFromPrev = (fv - p.LastVal) / float64(dp.Timestamp-p.Timestamp)
```

Comment:
Actually, we already verified that `p.timestamp < dp.Timestamp`. therefore `dp.Timestamp-p.Timestamp` will always be greater than zero.